File: php/testcode/redirect_010.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_redirect_double_encode
function redirect010(BenchmarkRequest $req): BenchmarkResponse {
    $url = urldecode($req->param('url'));
    header("Location: " . $url); // vuln-code-snippet vuln-line php_redirect_double_encode
    return BenchmarkResponse::ok('');
}
// vuln-code-snippet end php_redirect_double_encode

File: rust/testcode/redirect_010.rs
//! CWE-601: User-controlled path segment in URL allowing ../ traversal to external host.

// vuln-code-snippet start testcodeRedirect010
pub fn handle(req: &super::shared::BenchmarkRequest) -> super::shared::BenchmarkResponse {
    let user_path = req.param("path");
    let redirect = format!("/app/{}", user_path); // vuln-code-snippet target-line testcodeRedirect010
    super::shared::BenchmarkResponse::ok(&format!("Location: {}", redirect))
}
// vuln-code-snippet end testcodeRedirect010
